README.md
# Deep Learning Tutorials using PyTorch (PseudoLab)

[![gitlocalized ](https://gitlocalize.com/repo/5678/en/badge.svg)](https://gitlocalize.com/repo/5678/) [![License: CC BY-NC-SA 4.0](https://img.shields.io/badge/License-CC%20BY--NC--SA%204.0-lightgrey.svg)](https://creativecommons.org/licenses/by-nc-sa/4.0/)

English version of [Deep learning tutorials using Pytorch](https://pseudo-lab.github.io/Tutorial-Book/) (maintained by [PseudoLab](https://pseudo-lab.com/))

Currently being translated using [Gitlocalize](https://gitlocalize.com/repo/5678)

Will be launched as a website around March 2021


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

#### Release: Tutorial-Book-en 0.2.0 | Release Date: Mar. 26, 2021 (ADD OD TRANSLATION)

- add object detection tutorial translation 
  <br/><br/><br/>

#### Release: Tutorial-Book-en 0.1.1 | Release Date: Mar. 8, 2021 (EDIT LANDING PAGE AND TS)

- fix typo on Lindsay
- delete youtube videos on Time Series chapter
- edt translation moderator to translation editors on landing page
- add button to translate to korean and add license button 
  <br/><br/><br/>

#### Release: Tutorial-Book-en 0.1.0 | Release Date: Mar. 3, 2021 (INITIALIZE)

- initialized English website with time-series tutorial
  <br/><br/><br/>



book/_build/jupyter_execute/chapters/en/time-series/Ch2-EDA.py
# 2. EDA

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/Pseudo-Lab/Tutorial-Book/blob/master/book/chapters/time-series/Ch2-EDA.ipynb)

from IPython.display import YouTubeVideo

YouTubeVideo(id="9E130rYSzaE", width=560, height=315)


In the previous chapter, we examined the structure of deep learning models and evaluation metrics for time series analysis. In this chapter, we will explore and visualize the datasets we will be using to perform time series forecasting in the next chapter.

We will use COVID-19 daily case datasets for the time series forecasting. The datasets we are going to use are [COVID-19 Data Repository by the Center for Systems Science and Engineering (CSSE) at Johns Hopkins University](https://github.com/CSSEGISandData/COVID-19) and [Kaggle: Novel Corona Virus 2019 Dataset](https://www.kaggle.com/sudalairajkumar/novel-corona-virus-2019-dataset?select=covid_19_data.csv).

COVID-19 daily cases by country are updated in the repository of Johns Hopkins University every day, along with various types of datasets, which are provided by Johns Hopkins University on its website and Kaggle.

In chapter 2.1, we will download the COVID-19 daily case datasets to see what features are included in them. In 2.2, we will perform EDA (Exploratory Data Analysis) on the datasets, focusing on world countries, and then in 2.3, we will dive into the COVID-19 daily cases in South Korea.

## 2.1 Download Datasets

Firstly, we will download the COVID-19 daily case datasets. Using the data-loader function coded by Pseudo-Lab, we can easily download the datasets.

Download the datasets from the repository of Tutorial-Book-Utils on GitHub by using `git clone` to save them in the Colab enviroment.

!git clone https://github.com/Pseudo-Lab/Tutorial-Book-Utils

![dataset example](https://github.com/Pseudo-Lab/Tutorial-Book/blob/master/book/pics/TS-ch2img01.PNG?raw=true)

- Figure 2-1. Downloaded Files in Colab

In Figure 2-1, you can see `PL_data_loader.py` in the directory, in addition to `README.md` and `utils_ObjectDetection.py`. <code>PL_data_loader.py</code> contains the functions to download Google Drive datasets. It is possible to download the COVID-19 daily case datasets with the input of <code>COVIDTimeSeries</code> if we use the parameter <code>--data</code>.


!python Tutorial-Book-Utils/PL_data_loader.py --data COVIDTimeSeries

Next, let's unarchive the downloaded file by using `unzip`, which is a Linux command. Unnecessary outputs are not printed if we use the command `-q`.

!unzip -q COVIDTimeSeries.zip

![dataset example](https://github.com/Pseudo-Lab/Tutorial-Book/blob/master/book/pics/TS-ch2img02.PNG?raw=true)

- Figure 2-2. Downloaded Datasets

Once the `COVIDTimesSeries.zip` is unarchived, you can see two downloaded files: `covid_19_data.csv` and <code>time_series_covid19_confirmed_global.csv</code> (Figure 2-2).

We will visualize the global COVID-19 daily cases with `covid_19_data.csv` based on time series, and we will visualize the daily cases of COVID-19 in South Korea with `time_series_covid19_confirmed_global.csv`. Let's first view the saved values after loading each dataset.

import pandas as pd

all = pd.read_csv('covid_19_data.csv')
confirmed = pd.read_csv('time_series_covid19_confirmed_global.csv')

`all` contains daily confirmed cases, daily deaths, and daily recovered data. The dataframe of the dataset consists of `ObservationDate`, `Province/State`, `Country/Region`, `Confirmed`, `Deaths`, and `Recovered`.

all

`confirmed` is sequential data showing daily cases by country. `Country/Region` and `Province/State` are location data. `Long` and `Lat` are longitude and latitude, respectively. `MM/DD/YYYY` column records daily confirmed cases. The dataframe looks like below.

confirmed

## 2.2 EDA of World Data

Let's visualize the global COVID-19 daily cases using `all`. As mentioned above, there are two location data: `Province/State` and `Country/Region`. We will add the daily cases of every country based on `Country/Region`.


group = all.groupby(['ObservationDate', 'Country/Region'])['Confirmed'].sum()
group = group.reset_index()
group.head()

You can now see daily confirmed cases by country. Next, we will visualize world daily cases on the world map with animation effects. We can create the visualization with the `plotly.express` package. 

import plotly as py
import plotly.express as px

With `px.choropleth`, we create the map layer, and then update dates through `.update_layout`.

Each parameter is as follows:

- location : name of the column which indicates location information in the dataframe

- locationmode : the range of countries (‘ISO-3’, ‘USA-states’, and ‘country names’)

- color : name of the column which has data to be expressed on the diagram

- animation_frame : name of the column for animation effects

Press the play button to see the COVID-19 daily case graphs.

choro_map=px.choropleth(group, 
                    locations="Country/Region", 
                    locationmode = "country names",
                    color="Confirmed", 
                    animation_frame="ObservationDate"
                   )

choro_map.update_layout(
    title_text = 'Global Spread of Coronavirus',
    title_x = 0.5,
    geo=dict(
        showframe = False,
        showcoastlines = False,
    ))
    
choro_map.show()

You can see that COVID-19 begins in China and then gradually spreads throughout the world. According to the visualization, recently there are a lot of cumulative cases in North America, South America, and India, compared to other countries. 

## 2.3 EDA of South Korean Dataset

This time, we will extract the South Korean data from `confirmed`. The data is cumulative, which means it shows the sum of all cases on a given date.

confirmed[confirmed['Country/Region']=='Korea, South']

Excluding the `Lat` and `Long` columns, we only have the date and cases columns. For coding convenience, let's flip the rows and columns and change the date format from <code>str</code> to <code>datetime</code> (<code>to_datetime</code>).

korea = confirmed[confirmed['Country/Region']=='Korea, South'].iloc[:,4:].T
korea.index = pd.to_datetime(korea.index)
korea

Let's visualize the data with `matplotlib.pyplot` and `seaborn`, which are representative visualization packages. With `%matplotlib inline`, the matplotlib graphs will be included in the notebook, along with the code. You can change the figure size through `rcParams['figure.figsize']` from `pylab` and the style and font size through `sns.set`.

import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline
from pylab import rcParams
rcParams['figure.figsize'] = 12, 8
sns.set(style='whitegrid', palette='muted', font_scale=1.2)

plt.plot(korea)
plt.show()

Now, let's look at daily cases instead of cumulative cases. Using the `diff` function, it is possible to tell the difference between each row, which makes it easier to transform cumulative data into daily data. However, there is a missing value in the first row of the cumulative data, which we need to fill. Also, we will change the data format to an integer (<code>int</code>).

daily_cases = korea.diff().fillna(korea.iloc[0]).astype('int')
daily_cases

The following visualization is the output for daily cases. Since we already created a visualization, the visualization parameters will be based on the previous data unless you reset the parameters.

plt.plot(daily_cases)
plt.show()

There are spikes in the beginning of March and late August. At the end of the year, new daily cases skyrocket.

In this chapter, we explored COVID-19 confirmed case datasets in South Korea and across the world. In the next chapter, we will learn about data pre-processing of COVID-19 daily cases, using South Korea as our example, and use a deep learning model for time series forecasting.

book/_build/jupyter_execute/chapters/en/time-series/Ch3-preprocessing.py
# 3. Data Pre-Processing

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/Pseudo-Lab/Tutorial-Book/blob/master/book/chapters/time-series/Ch3-preprocessing.ipynb)

from IPython.display import HTML

HTML('<iframe width="560" height="315" src="https://www.youtube.com/embed/mgkUDA-V9oA" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>')


In the previous chapter, we explored dataset features through an EDA. In this chapter, we will learn how to pre-process data for a time series task.

Pre-processing data into pairs of features and target variable is required in order to use a sequential dataset for supervised learning. In addition, it is necessary to unify data scales to stably train deep learning models. In chapter 3.1, we will transform the raw data of COVID-19 confirmed cases into data for supervised learning, and in chapter 3.2, we will examine how to perform data scaling. 

## 3.1 Preparing Data for Supervised Learning

We will load a dataset for data pre-processing, using code introduced in chapter 2.1.

!git clone https://github.com/Pseudo-Lab/Tutorial-Book-Utils
!python Tutorial-Book-Utils/PL_data_loader.py --data COVIDTimeSeries
!unzip -q COVIDTimeSeries.zip

Let's extract `daily_cases`, which shows the daily confirmed COVID-19 cases for South Korea, using code introduced in chapter 2.3.

import pandas as pd
confirmed = pd.read_csv('time_series_covid19_confirmed_global.csv')
confirmed[confirmed['Country/Region']=='Korea, South']
korea = confirmed[confirmed['Country/Region']=='Korea, South'].iloc[:,4:].T
korea.index = pd.to_datetime(korea.index)
daily_cases = korea.diff().fillna(korea.iloc[0]).astype('int')
daily_cases

We need to convert the time series data shown above into pairs of input and output variables to use them for supervised learning. In a time series task, we call this kind of data as sequential data. Firstly, we need to define the sequence length in order to transform the data into sequential data. The sequence length is decided by how many days from the data we wish to use to predict future cases. For example, for a sequence length of 5, data in $t-1$, $t-2$, $t-3$, $t-4$, and $t-5$ are used to predict data in time $t$. Likewise, a task where we predict variable at time $t$ using data from $t-k$ to $t-1$ is called an one-step prediction task.

The `create_sequences` function defined below transforms time series data with size `N` into data with a <code>N - seq-length</code> size for supervised learning (See Figure 3-1).

![](https://github.com/Pseudo-Lab/Tutorial-Book/blob/master/book/pics/TS-ch3img01.png?raw=true)


- Figure 3-1 Transforming Process of Times Series Data

import numpy as np

def create_sequences(data, seq_length):
    xs = []
    ys = []
    for i in range(len(data)-seq_length):
        x = data.iloc[i:(i+seq_length)]
        y = data.iloc[i+seq_length]
        xs.append(x)
        ys.append(y)
    return np.array(xs), np.array(ys)

seq_length = 5
X, y = create_sequences(daily_cases, seq_length)

With `seq_length` = 5, applying the `create_sequences` function to `daily_cases`, we got 327 samples in total for supervised learning. 

X.shape, y.shape

We will divide the transformed dataset into training, validation, and test datasets with an 8:1:1 ratio. The total number of data is 327, so the division of each dataset results in the following: 261 data for training, 33 data for validation, and 33 data for testing.

train_size = int(327 * 0.8)
print(train_size)

X_train, y_train = X[:train_size], y[:train_size]
X_val, y_val = X[train_size:train_size+33], y[train_size:train_size+33]
X_test, y_test = X[train_size+33:], y[train_size+33:]

print(X_train.shape, X_val.shape, X_test.shape)
print(y_train.shape, y_val.shape, y_test.shape)

## 3.2 Data Scaling

In chapter 3.2, we will perform data scaling. More specifically, we will perform MinMax scaling, which transforms the data range to between 0 and 1. Apply the following mathematical notation for MinMax scaling after calculating the minimum and maximum values of the data group.


> $x_{scaled} = \displaystyle\frac{x - x_{min}}{x_{max} - x_{min}}$

The data scaling for the training, validation, and test datasets must be processed based on the statistics of the training data. Input variables from the testing dataset should not be used, so we must perform training dataset scaling using the statitistics of the training data.<br>Since the model was trained with the statistics of the training data, the test data must also be scaled based on the same values in order to evaluate model performance later. Similarly, the validation data require data scaling based on the statistics of the training data, since validation data need to go through the same process of pre-processing as the test data.

We will get the minimum and maximum values from the `X_train` data in order to apply MinMax scaling.

MIN = X_train.min()
MAX = X_train.max()
print(MIN, MAX)

The minimum and maximum values are 0 and 851, respectively. Next, we will define the MinMax scaling function. 

def MinMaxScale(array, min, max):

    return (array - min) / (max - min)

Let's perform scaling using the `MinMaxScale` function.

X_train = MinMaxScale(X_train, MIN, MAX)
y_train = MinMaxScale(y_train, MIN, MAX)
X_val = MinMaxScale(X_val, MIN, MAX)
y_val = MinMaxScale(y_val, MIN, MAX)
X_test = MinMaxScale(X_test, MIN, MAX)
y_test = MinMaxScale(y_test, MIN, MAX)

Next, we will transform the data type from `np.array` into `torch.Tensor` in order for the data to be input in a PyTorch model. First, we will define the function for transforming the data type. 

import torch

def make_Tensor(array):
    return torch.from_numpy(array).float()

We will perform the transformation through the `make_Tensor` function.

X_train = make_Tensor(X_train)
y_train = make_Tensor(y_train)
X_val = make_Tensor(X_val)
y_val = make_Tensor(y_val)
X_test = make_Tensor(X_test)
y_test = make_Tensor(y_test)

So far, we have practiced transforming data into the correct format for the supervised learning of time series and data scaling. In the next chapter, we will build a prediction model for COVID-19 cases with the data we curated. 

book/_build/jupyter_execute/chapters/en/time-series/Ch1-Time-Series.py
# 1. Introduction to Time Series

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/Pseudo-Lab/Tutorial-Book/blob/master/book/chapters/time-series/Ch1-Time-Series.ipynb)

from IPython.display import HTML

HTML('<iframe width="560" height="315" src="https://www.youtube.com/embed/DRZFhCBsGQY" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>')


Time series prediction is the prediction of future values based on observed values in the past. It can be defined as a supervised learning problem in that it requires finding patterns between observed data in the past and future values. Therefore, in this chapter, we will build a model that predicts future values through supervised learning based on a neural network structure.

Time series prediction is a skill that is required in many areas. One of the areas is in the energy domain. Electric power plants need to predict future power demand to secure sufficient amounts of reserved power, and city gas companies need a future usage prediction model to take preemptive measures against meter reader failure and meter reader cheating. In fact, these issues were presented for data science competitions ([electric power](https://dacon.io/competitions/official/235606/overview/), [city gas](https://icim.nims.re.kr/platform/question/16)) to facilitate the discovery of new models. In addition, stakeholders in the retail domain are interested in predicting the sales volumes of items for efficient product management, which was also the topic of a data science competition.([distribution](https://www.kaggle.com/c/m5-forecasting-accuracy/overview))

In this tutorial, we will build a model that predicts future confirmed cases of COVID-19 based on data from the past confirmed cases using [COVID-19 confirmed case data](https://github.com/CSSEGISandData/COVID-19) provided by Johns Hopkins University's Center for Systems Science and Engineering. In chapter 1, we will look at the neural network structures that can be used when building a time series prediction model, and check the metrics that can be used when evaluating model performance. In chapter 2, we will deepen our understanding of COVID-19 confirmed case data through exploratory data analysis, and in chapter 3 we will learn how to restructure time series data so it can be used for supervised learning. In chapters 4 and 5, we will use deep learning models to predict future confirmed cases.


## 1.1 Available Deep Learning Architectures

### 1.1.1 CNN

<p align="center"> <img align="center" src="https://github.com/Pseudo-Lab/Tutorial-Book/blob/master/book/pics/TS-ch1img01.PNG?raw=true"></p>


- Figure 1-1. Example of CNN structure (Reference: Lim et al. 2020. Time Series Forecasting With Deep Learning: A Survey)


In general, CNN is a network structure that performs well in computer vision tasks. However, CNNs can also be applied to time series prediction. A weighted sum between input sequence data can be calculated using a one-dimensional convolution filter to calculate the predicted future value. However, the CNN structure does not take into account the temporal dependence between past and future data. 

### 1.1.2 RNN

<p align="center"> <img align="center" src="https://github.com/Pseudo-Lab/Tutorial-Book/blob/master/book/pics/TS-ch1img02.PNG?raw=true"></p>


- Figure 1-2. Example of RNN Structure (Reference: Lim et al. 2020. Time Series Forecasting With Deep Learning: A Survey) 

RNN is a network structure that is frequently used in solving natural language processing tasks, and it utilizes hidden state information accumulated in previous states for future prediction. By doing so, it is possible to manipulate past information to calculate future forecasts. However, if the given input sequence is too large, a vanishing gradient problem may occur that adversely affects model training. Therefore, the LSTM structure, which solves the vanishing gradient problem, is used frequently. We will use the LSTM structure in this tutorial. 

### 1.1.3 Attention Mechanism

<p align="center"> <img align="center" src="https://github.com/Pseudo-Lab/Tutorial-Book/blob/master/book/pics/TS-ch1img03.PNG?raw=true"></p>


- Figure 1-3. Example of Attention Mechanism (Reference: Lim et al. 2020. Time Series Forecasting With Deep Learning: A Survey) 

When predicting the future based on past information, there can be information from the past that is helpful and information that isn't. For example, if a retailer wants to predict weekend sales, it may be helpful to consider weekend sales from the previous week rather than sales on the weekdays leading up to the prediction period. By using the attention mechanism, it is possible to manipulate these factors into the model. It calculates the importance of past points in time and uses them to inference future values. More accurate prediction is possible by assigning more weight to the value that is directly influencing the point to be predicted. 

## 1.2 Evaluation Metric

In this tutorial, we are going to build a model for predicting confirmed cases of COVID-19. Since the confirmed cases are measured with continuous values, the performance of the model can be evaluated through the gaps between the predicted and actual values. In this section, we'll look at various ways to calculate the gaps between the predicted and actual values. We will first define some symbols before diving in.


> $y_i$: Actual values that need to be predicted <br> $\hat{y}_i$: Values predicted by the model <br> $n$: Size of the test dataset


Chapters 1.2.1 to 1.2.4 will use the symbols above, but the definition of the symbols will change in chapter 1.2.5. Please be aware of this change.

### 1.2.1 MAE (Mean Absolute Error)

> $MAE=\frac{1}{n}\displaystyle\sum_{i=1}^{n} |y_i-\hat{y}_i|$

MAE, also known as L1 Loss, can be calculated by dividing the sum of the absolute differences between the predicted values and the actual values by the number of samples(n). Since this is the process for calculating an average, from now on we will refer to this as 'calculating the mean'. Since the scale of MAE is the same scale as the target variable being predicted, the meaning of the value can be understood intuitively. The implemented code looks like this: 

import numpy as np #import numpy package

def MAE(true, pred):
    '''
    true: np.array 
    pred: np.array
    '''
    return np.mean(np.abs(true-pred))

TRUE = np.array([10, 20, 30, 40, 50])
PRED = np.array([30, 40, 50, 60, 70])

MAE(TRUE, PRED)

### 1.2.2 MSE (Mean Squared Error)

> $MSE=\frac{1}{n}\displaystyle\sum_{i=1}^{n} (y_i-\hat{y}_i)^2$

> $RMSE=\sqrt{\frac{1}{n}\displaystyle\sum_{i=1}^{n} (y_i-\hat{y}_i)^2}$


MSE, also known as L2 Loss, is calculated by taking the mean of the squared differences between the predicted values and the actual values. The more the predicted values deviate from the actual values, the more the MSE value will increase. It will increase exponentially. Since the calculated value is squared, the scale of the target variable and MSE is different. In order to match the scale of the target value, we need to calculate the square root of the MSE. This value is called RMSE. The implemented code for MSE looks like this:

def MSE(true, pred):
    '''
    true: np.array 
    pred: np.array
    '''
    return np.mean(np.square(true-pred))

TRUE = np.array([10, 20, 30, 40, 50])
PRED = np.array([30, 40, 50, 60, 70])

MSE(TRUE, PRED)

### 1.2.3 MAPE (Mean Absolute Percentage Error)

> $MAPE=\frac{1}{n}\displaystyle\sum_{i=1}^{n} |\frac{y_i-\hat{y}_i}{y_i}|$


(Reference: https://en.wikipedia.org/wiki/Mean_absolute_percentage_error)

In order to calculate MAPE, first calculate the relative size of the error compared to the actual values by dividing the difference between each of the actual values and the predicted value by each actual value. Then, take the absolute value of the relative size of the error for each actual value and calculate the mean. Since the size of the error is expressed as a percentage value, it can be used to understand the performance of the model. Also, it is a suitable metric for evaluating a model when there is more than one target variable because the scale of the calculated errors across the target variables will be similar.

However, if there is an actual value of 0, MAPE will be undefined. In addition, even if the absolute values of the errors are same, more penalties are added to a predicted value that overestimates([Makridakis, 1993](https://doi.org/10.1016/0169-2070(93)90079-3)). Let's look at the example below. 

def MAPE(true, pred):
    '''
    true: np.array 
    pred: np.array
    '''
    return np.mean(np.abs((true-pred)/true))

TRUE_UNDER = np.array([10, 20, 30, 40, 50])
PRED_OVER = np.array([30, 40, 50, 60, 70])
TRUE_OVER = np.array([30, 40, 50, 60, 70])
PRED_UNDER = np.array([10, 20, 30, 40, 50])


print('Comparison between MAE, MAPE when average error is 20 depending on the relationship between actual and predicted value \n')

print('When actual value is smaller than predicted value (Overestimating)')
print('MAE:', MAE(TRUE_UNDER, PRED_OVER))
print('MAPE:', MAPE(TRUE_UNDER, PRED_OVER))


print('\nWhen actual value is bigger than predicted value (Underestimating)')
print('MAE:', MAE(TRUE_OVER, PRED_UNDER))
print('MAPE:', MAPE(TRUE_OVER, PRED_UNDER))


MAPE divides the error by the actual value $y$ to convert it to a percentage. Therefore, the calculated value is dependent on $y$. Even if the numerators are the same, smaller denominators will increase the overall error.

We can observe this phenomenon by observing the two examples above where (`TRUE_UNDER`, `PRED_OVER`) predicts values that are more than the actual values by 20 and (`TRUE_OVER`, `PRED_UNDER`) predicts values that are less than the actual values by 20. On both examples, the MAE values are the same at 20. However, for the `TRUE_UNDER` case the MAPE value is calculated as 0.913 and `TRUE_OVER` case calculates the MAPE value as 0.437. 

### 1.2.4 SMAPE (Symmetric Mean Absolute Percentage Error)


> $SMAPE=\frac{100}{n}\displaystyle\sum_{i=1}^{n} \frac{|y_i-\hat{y}_i|}{|y_i| + |\hat{y}_i|}$


(Reference: https://en.wikipedia.org/wiki/Symmetric_mean_absolute_percentage_error)


SMAPE has been created to deal with the limitations of MAPE for the examples above([Makridakis, 1993](https://doi.org/10.1016/0169-2070(93)90079-3)). Let's look at the example below.

def SMAPE(true, pred):
    '''
    true: np.array 
    pred: np.array
    '''
    return np.mean((np.abs(true-pred))/(np.abs(true) + np.abs(pred))) #we won't include 100 in this code since it's a constant

print('Comparison between MAE, SMAPE when average error is 20 \n')

print('When actual value is smaller than predicted value (Overestimating)')
print('MAE:', MAE(TRUE_UNDER, PRED_OVER))
print('SMAPE:', SMAPE(TRUE_UNDER, PRED_OVER))


print('\nWhen actual value is bigger than predicted value (Underestimating)')
print('MAE:', MAE(TRUE_OVER, PRED_UNDER))
print('SMAPE:', SMAPE(TRUE_OVER, PRED_UNDER))


We can observe that MAPE produced different values of 0.91 and 0.43 respectively on the same example, but SMAPE yielded the same values of 0.29. However, SMAPE is dependent on $\hat{y}_i$ because the predicted value $\hat{y}_i$ is included in the denominator. When the predicted value is an underestimation, the denominator becomes smaller and the overall error increases. Let's look at the example below. 

TRUE2 = np.array([40, 50, 60, 70, 80])
PRED2_UNDER = np.array([20, 30, 40, 50, 60])
PRED2_OVER = np.array([60, 70, 80, 90, 100])

print('Comparison between MAE, SMAPE when average error is 20 \n')

print('When overestimating')
print('MAE:', MAE(TRUE2, PRED2_OVER))
print('SMAPE:', SMAPE(TRUE2, PRED2_OVER))

print('\nWhen underestimating')
print('MAE:', MAE(TRUE2, PRED2_UNDER))
print('SMAPE:', SMAPE(TRUE2, PRED2_UNDER))

`PRED2_UNDER` and `PRED2_OVER` both have an MAE of 20 compared with `TRUE2` , but SMAPE is calculated as 0.218 for `PRED2_UNDER` where underestimation occurred and 0.149 for `PRED2_OVER` where overestimation occurred.

### 1.2.5 RMSSE (Root Mean Squared Scaled Error)

> $RMSSE=\sqrt{\displaystyle\frac{\frac{1}{h}\sum_{i=n+1}^{n+h} (y_i-\hat{y}*i)^2}{\frac{1}{n-1}\sum*{i=2}^{n} (y_i-y_{i-1})^2}}$

We will proceed with the definition of the symbols used in the RMSSE formula. Each symbol has the following meaning.

> $y_i$: Actual value to be predicted
>
> $\hat{y}_i$: Value predicted by the model
>
> $n$: Size of the training dataset
>
> $h$: Size of the test dataset

RMSSE is a modified form of Mean Absolute Scaled Error ( [Hyndman, 2006](https://doi.org/10.1016/j.ijforecast.2006.03.001) ) and solves the problems of MAPE and SMAPE mentioned above. We have seen from above examples that MAPE and SMAPE result in an uneven overall error depending on the underestimation or overestimation of the model since they use the actual and predicted values of the test data to scale the MAE.

RMSSE avoids this problem by using the training data when scaling the MSE. It scales the error by dividing by the MSE that is calculated when applying naive forecasting on the training data, so the overall error is not affected by the underestimation or overestimation of the model. The naive forecast method is a method of forecasting from the most recent observation and is defined as follows:

> $\hat{y}*i = y*{i-1}$

This is a method of predicting the value at the time of $i$ as the actual value at the time of $i-1$. Since it is divided by the MSE value for the naive forecast method, if the RMSSE value is greater than 1, it means that the overall performance of the model is not better than the naive forecast method. On the other hand, if RMSSE is less than 1, it means that the performance of the model is better than the naive forecast method. Let's implement RMSSE with the code below. 

def RMSSE(true, pred, train): 
    '''
    true: np.array 
    pred: np.array
    train: np.array
    '''
    
    n = len(train)

    numerator = np.mean(np.sum(np.square(true - pred)))
    
    denominator = 1/(n-1)*np.sum(np.square((train[1:] - train[:-1])))
    
    msse = numerator/denominator
    
    return msse ** 0.5

TRAIN = np.array([10, 20, 30, 40, 50]) #create a random training dataset for calculating RMSSE

print(RMSSE(TRUE_UNDER, PRED_OVER, TRAIN))
print(RMSSE(TRUE_OVER, PRED_UNDER, TRAIN))
print(RMSSE(TRUE2, PRED2_OVER, TRAIN))
print(RMSSE(TRUE2, PRED2_UNDER, TRAIN))

On the examples where MAPE and SMAPE gave unequal penalties even if the MAE values were the same, RMSSE not only gave equal penalties, but also scaled the error.

So far, we have looked at the deep learning network structures and evaluation metrics that can be used for evaluating time series forecasting models. In the next chapter, we will perform exploratory data analysis on the COVID-19 data which will be used for building a forecasting model. 